common/src/main/java/io/pravega/common/util/BufferView.java

```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+/**
+ * Defines a generic read-only view of a readable memory buffer with a known length.
+ * For array-backed Buffers, see {@link ArrayView}.
+ */
+public interface BufferView {
+    /**
+     * Gets a value representing the length of this {@link BufferView}.
+     *
+     * @return The length.
+     */
+    int getLength();
+
+    /**
+     * Creates an InputStream that can be used to read the contents of this {@link BufferView}. The InputStream returned
+     * spans the entire {@link BufferView}.
+     *
+     * @return The InputStream.
+     */
+    InputStream getReader();
+
+    /**
+     * Returns a copy of the contents of this {@link BufferView}.
+     *
+     * @return A byte array with the same length as this ArrayView, containing a copy of the data within it.
+     */
+    byte[] getCopy();
+
+    /**
+     * Copies the contents of this {@link BufferView} to the given {@link OutputStream} using a 4KB copy buffer.
+     *
+     * @param target The {@link OutputStream} to write to.
+     * @throws IOException If an exception occurred.
+     */
+    void copyTo(OutputStream target) throws IOException;
+
+    /**
+     * When implemented in a derived class, notifies any wrapped buffer that this {@link BufferView} has a need for it.
+     * Use {@link #release()} to do the opposite.
+     */
+    default void retain() {
```

Comment:
I actually had it like that before, but it's not so simple, and let me explain why.

On the append path in the Segment Store:
1. We create a `BufferView/ByteBufWrapper` in `AppendProcessor` which we pass down to the Segment Store.
2. The Segment Store may reject the request synchronously, or asynchronously (for various reasons). In each case it would throw an exception or return a failed CompletableFuture.
3. The Segment Store may accept the request, write it to Tier 1, ack it back to AppendProcessor and the asynchronously add it to the Cache.

We need to handle all these 3 cases. If we have `ByteBufWrapper/BufferView` implement AutoCloseable, then we could add some code to `AppendProcessor` (which creates it initially) to close it once the returned Future from the Store completes. Actually this is what `AppendProcessor` does anyway: it invokes `release` on the underlying `ByteBuf` after the future completes (which is correct, since `ByteBuf` sets the refCnt to 1 when it is created).

But that may be premature! In case 3 above, the BufferView may still be needed so it can be written to the cache when `AppendProcessor` invokes `release`, which means we won't be able to use it anymore.

As such, I decided the best way to handle this is the following:
- Have `retain` and `release`. Make them optional. Simply creating a `ByteBufWrapper` will not touch the underlying `ByteBuf` refCnt.
- If we end up in cases 1 or 2, all is good, since we haven't increased the refCnt.
- In case 3, when we create a `StreamSegmentAppendOperation` (which has a payload - `BufferView`), that class' constructor will invoke `retain` and I made that class `AutoCloseable`, in which case it will release the buffer.
    - The Segment Store will invoke `close` on that operation when it is completely done with it, which will ensure the buffer is released. It will also call `close` if it has already created it but had to abort the append for whatever reason.